packages/coding-standard/src/Rules/AbstractPrefferedCallOverFuncRule.php

```diff
@@ -12,6 +12,16 @@
 
 abstract class AbstractPrefferedCallOverFuncRule extends AbstractSymplifyRule
 {
+    /**
+     * @var string
+     */
+    public const PREFER_STATIC_CALL_ERROR_MESSAGE = 'Use "%s::%s()" static call over "%s()" func call';
+
+    /**
+     * @var string
+     */
+    public const PREFER_METHOD_CALL_ERROR_MESSAGE = 'Use "%s::%s()" method call over "%s()" func call';
+
```

Comment:
This is actually inverse coupling to the child. The parent should now know anything about their children.

If new child would be added, the parent cannot be changed with extra constants.

How could we decouple it back to previous form?

Reply:
previously, I used error message in child, but got phpstan error as parent can't find the constant, any idea?

Reply:
I moved back error message to child classes

Reply:
it seems we can use value object's contants for it ref https://3v4l.org/aNSte

Reply:
> previously, I used error message in child, but got phpstan error as parent can't find the constant, any idea?

That is correct. The parent should not know about it's child constants.

> it seems we can use value object's contants for it ref 3v4l.org/aNSte

The message should be in particular rule by convention. If we move it to another class, the convention is broken = contributors will now have 2 clues about where to put them = mess.

---

What about using check method in abstract with bool return and handling the return of error message in a child?

Reply:
currently, it already in child, and in parent, use try cath: 

```php
            try {
                $constantReflex = new ReflectionClassConstant(static::class, 'ERROR_MESSAGE');
                $errorMessage = $constantReflex->getValue();
            } catch (ReflectionException $reflectionException) {
                throw new ShouldNotHappenException('const ERROR_MESSAGE must be defined with public modifier');
            }
```

to handle phpstan notice.

Do you mean add function `getErrorMessage()` ?

Reply:
Yes, it seems we can use function `isSkipped()` in abstract class, and the `process()` function itself is in the child.

Reply:
method that return bool in abstract class seems not possible as the check is in the loop with message generated as : 

```php
            $errorMessage = sprintf($errorMessage, $staticCall[0], $staticCall[1], $funcCall);
            return [$errorMessage];
```

which variables came from the loop

Reply:
I refactored to have `getErrorMessageParameters()` in abstract class, use it to generate error message in child